embassy-extras/src/fmt.rs

```diff
@@ -1,67 +1,168 @@
+            #[cfg(feature = "defmt")]
+            defmt::assert_eq!($($x)*);
+        }
     };
 }
 
-#[cfg(feature = "log")]
-mod fmt {
-    pub use core::{
-        assert, assert_eq, assert_ne, debug_assert, debug_assert_eq, debug_assert_ne, panic, todo,
-        unreachable,
+macro_rules! assert_ne {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::assert_ne!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::assert_ne!($($x)*);
+        }
     };
-    pub use log::{debug, error, info, trace, warn};
 }
 
-#[cfg(not(any(feature = "defmt", feature = "log")))]
-mod fmt {
-    #![macro_use]
+macro_rules! debug_assert {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::debug_assert!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::debug_assert!($($x)*);
+        }
+    };
+}
 
-    pub use core::{
-        assert, assert_eq, assert_ne, debug_assert, debug_assert_eq, debug_assert_ne, panic, todo,
-        unreachable,
+macro_rules! debug_assert_eq {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::debug_assert_eq!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::debug_assert_eq!($($x)*);
+        }
     };
+}
 
-    macro_rules! trace {
-        ($($msg:expr),+ $(,)?) => {
-            ()
-        };
-    }
+macro_rules! debug_assert_ne {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::debug_assert_ne!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::debug_assert_ne!($($x)*);
+        }
+    };
+}
 
-    macro_rules! debug {
-        ($($msg:expr),+ $(,)?) => {
-            ()
-        };
-    }
+macro_rules! todo {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::todo!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::todo!($($x)*);
+        }
+    };
+}
 
-    macro_rules! info {
-        ($($msg:expr),+ $(,)?) => {
-            ()
-        };
-    }
+macro_rules! unreachable {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::unreachable!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::unreachable!($($x)*);
+        }
+    };
+}
 
-    macro_rules! warn {
-        ($($msg:expr),+ $(,)?) => {
-            ()
-        };
-    }
+macro_rules! panic {
+    ($($x:tt)*) => {
+        {
+            #[cfg(not(feature = "defmt"))]
+            core::panic!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::panic!($($x)*);
+        }
+    };
+}
 
-    macro_rules! error {
-        ($($msg:expr),+ $(,)?) => {
-            ()
-        };
-    }
+macro_rules! trace {
+    ($($x:tt)*) => {
+        {
+            #[cfg(feature = "log")]
+            log::trace!($($x)*);
+            #[cfg(feature = "defmt")]
+            defmt::trace!($($x)*);
```

Comment:
Hmm, will this produce an unused warning when compiling without defmt nor log ? Maybe do something like smoltcp:
```
$( let _ = $x; )*
```
But not sure how that works with `tt`.

Reply:
Done